pkg/kubelet/images/image_manager_test.go

```diff
@@ -201,7 +217,7 @@ func TestParallelPuller(t *testing.T) {
 
 	useSerializedEnv := false
 	for _, c := range cases {
-		puller, fakeClock, fakeRuntime, container := pullerTestEnv(c, useSerializedEnv)
+		puller, fakeClock, fakeRuntime, container, fakePodPullingTimeRecorder := pullerTestEnv(c, useSerializedEnv)
```

Comment:
I think this should go inside the `t.Run` so it starts clean.

I also think that this way you don't need the reset

Reply:
I changed to `reset()`. 

When moving `pullerTestEnv` inside the `t.Run` it panics on `(*FakeRuntime).assertList` due to an invalid memory address or nil pointer dereference. 
I wonder if `pullerTestEnv` was called outside of `t.Run` for this reason, or it might be due to the loop ran inside the `t.Run` and the reset of `fakeRuntime.CalledFunctions` to `nil`, as each subtest is run in a separate goroutine.

Reply:
do you have the entire stack trace?

Reply:
```
--- FAIL: TestSerializedPuller (0.00s)
    --- FAIL: TestSerializedPuller/image_missing,_pull_if_not_present0 (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x21c95b0]

goroutine 81 [running]:
testing.tRunner.func1.2({0x24c0c00, 0x3731810})
	/usr/local/Cellar/go/1.19.4/libexec/src/testing/testing.go:1396 +0x24e
testing.tRunner.func1()
	/usr/local/Cellar/go/1.19.4/libexec/src/testing/testing.go:1399 +0x39f
panic({0x24c0c00, 0x3731810})
	/usr/local/Cellar/go/1.19.4/libexec/src/runtime/panic.go:884 +0x212
k8s.io/kubernetes/pkg/kubelet/container/testing.(*FakeRuntime).assertList(0xc0005b4580, {0xc0001113d0, 0x1, 0x1}, {0xc00049e0e0, 0x2, 0x2})
	/Users/thomasmilox/perso/kubernetes/pkg/kubelet/container/testing/fake_runtime.go:141 +0x110
k8s.io/kubernetes/pkg/kubelet/container/testing.(*FakeRuntime).AssertCalls(0xc000116280?, {0xc0001113d0?, 0xc00004a0c0?, 0xc000600480?})
	/Users/thomasmilox/perso/kubernetes/pkg/kubelet/container/testing/fake_runtime.go:151 +0xc5
k8s.io/kubernetes/pkg/kubelet/images.TestSerializedPuller.func1(0x0?)
	/Users/thomasmilox/perso/kubernetes/pkg/kubelet/images/image_manager_test.go:260 +0x15c
testing.tRunner(0xc000103d40, 0xc00015e1e0)
	/usr/local/Cellar/go/1.19.4/libexec/src/testing/testing.go:1446 +0x10b
created by testing.(*T).Run
	/usr/local/Cellar/go/1.19.4/libexec/src/testing/testing.go:1493 +0x35f
FAIL	k8s.io/kubernetes/pkg/kubelet/images	0.568s
FAI
```

Reply:
The nil pointer dereference comes from `fakeRuntime.T` used in `assertList()` which is nil here.
I could rewrite the function like this to avoid the `reset()`:
```go
func TestSerializedPuller(t *testing.T) {
	pod := &v1.Pod{
		ObjectMeta: metav1.ObjectMeta{
			Name:            "test_pod",
			Namespace:       "test-ns",
			UID:             "bar",
			ResourceVersion: "42",
		}}

	cases := pullerTestCases()

	useSerializedEnv := true
	for _, c := range cases {
		for i, expected := range c.expected {
			t.Run(fmt.Sprintf("%s%d", c.testName, i), func(t *testing.T) {
				ctx := context.Background()
				puller, fakeClock, fakeRuntime, container, fakePodPullingTimeRecorder := pullerTestEnv(c, useSerializedEnv)
				fakeRuntime.CalledFunctions = nil
				fakeClock.Step(time.Second)
				fakeRuntime.T = t // just for debugging

				_, _, err := puller.EnsureImageExists(ctx, pod, container, nil, nil)
				fakeRuntime.AssertCalls(expected.calls)
				assert.Equal(t, expected.err, err)
				assert.Equal(t, expected.podPullingStartTimeShouldHaveBeenRecorded, fakePodPullingTimeRecorder.startedPullingRecorded)
				assert.Equal(t, expected.podPullingFinishTimeShouldHaveBeenRecorded, fakePodPullingTimeRecorder.finishedPullingRecorded)
			})
		}
	}
}
```
But then the test fails:
```
--- FAIL: TestSerializedPuller (0.00s)
    --- FAIL: TestSerializedPuller/image_missing,_unable_to_fetch2 (0.00s)
        /Users/thomasmilox/perso/kubernetes/pkg/kubelet/images/fake_runtime.go:141: AssertList: expected []string{"GetImageRef"}, got []string{"GetImageRef", "PullImage"}
        /Users/thomasmilox/perso/kubernetes/pkg/kubelet/images/image_manager_test.go:261: 
            	Error Trace:	/Users/thomasmilox/perso/kubernetes/pkg/kubelet/images/image_manager_test.go:261
            	Error:      	Not equal: 
            	            	expected: &errors.errorString{s:"ImagePullBackOff"}
            	            	actual  : &errors.errorString{s:"ErrImagePull"}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,3 +1,3 @@
            	            	 (*errors.errorString)({
            	            	- s: (string) (len=16) "ImagePullBackOff"
            	            	+ s: (string) (len=12) "ErrImagePull"
            	            	 })
            	Test:       	TestSerializedPuller/image_missing,_unable_to_fetch2
```

Reply:
The nil pointer dereference comes from here https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/container/testing/fake_runtime.go#L141
If I Add this to understand why the deep equality fails
```go
fakeRuntime.T = t
```
I got:
```
--- FAIL: TestParallelPuller (0.00s)
    --- FAIL: TestParallelPuller/image_missing,_unable_to_fetch (0.00s)
        /Users/thomasmilox/perso/kubernetes/pkg/kubelet/images/fake_runtime.go:141: AssertList: expected []string{"GetImageRef"}, got []string{"GetImageRef", "PullImage"}
        /Users/thomasmilox/perso/kubernetes/pkg/kubelet/images/image_manager_test.go:230: 
            	Error Trace:	/Users/thomasmilox/perso/kubernetes/pkg/kubelet/images/image_manager_test.go:230
            	Error:      	Not equal: 
            	            	expected: &errors.errorString{s:"ImagePullBackOff"}
            	            	actual  : &errors.errorString{s:"ErrImagePull"}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,3 +1,3 @@
            	            	 (*errors.errorString)({
            	            	- s: (string) (len=16) "ImagePullBackOff"
            	            	+ s: (string) (len=12) "ErrImagePull"
            	            	 })
```
I will continue looking into it tomorrow but the order of `pullerTestCase` has to matter. I am obviously missing something here

Reply:
it is best practice to have independent test cases , dependent test cases are harder to debug and read and easy to break

Reply:
Yes I agree, but I only focused on testing the tracking of the image pulling and didn't touch the existing scenarios. The outcome of some tests seems to be order dependant indeed. What would you recommend here ? staying with the reset handler ?

Reply:
reset() works for me

The constructor inside t.Run doesn't panic for me

```diff
diff --git a/pkg/kubelet/images/image_manager_test.go b/pkg/kubelet/images/image_manager_test.go
index 4886828ba60..1ac1fc2816a 100644
--- a/pkg/kubelet/images/image_manager_test.go
+++ b/pkg/kubelet/images/image_manager_test.go
@@ -217,9 +217,9 @@ func TestParallelPuller(t *testing.T) {
 
        useSerializedEnv := false
        for _, c := range cases {
-               puller, fakeClock, fakeRuntime, container, fakePodPullingTimeRecorder := pullerTestEnv(c, useSerializedEnv)
-
                t.Run(c.testName, func(t *testing.T) {
+                       puller, fakeClock, fakeRuntime, container, fakePodPullingTimeRecorder := pullerTestEnv(c, useSerializedEnv)
+
                        ctx := context.Background()
                        for _, expected := range c.expected {
                                fakeRuntime.CalledFunctions = nil
@@ -248,9 +248,10 @@ func TestSerializedPuller(t *testing.T) {
 
        useSerializedEnv := true
        for _, c := range cases {
-               puller, fakeClock, fakeRuntime, container, fakePodPullingTimeRecorder := pullerTestEnv(c, useSerializedEnv)
 
                t.Run(c.testName, func(t *testing.T) {
+                       puller, fakeClock, fakeRuntime, container, fakePodPullingTimeRecorder := pullerTestEnv(c, useSerializedEnv)
+
                        ctx := context.Background()
                        for _, expected := range c.expected {
                                fakeRuntime.CalledFunctions = nil
@@ -308,13 +309,12 @@ func TestPullAndListImageWithPodAnnotations(t *testing.T) {
                        {[]string{"GetImageRef", "PullImage"}, nil, true, true},
                }}
 
-       useSerializedEnv := true
-       puller, fakeClock, fakeRuntime, container, fakePodPullingTimeRecorder := pullerTestEnv(c, useSerializedEnv)
-       fakeRuntime.CalledFunctions = nil
-       fakeRuntime.ImageList = []Image{}
-       fakeClock.Step(time.Second)
-
        t.Run(c.testName, func(t *testing.T) {
+               useSerializedEnv := true
+               puller, fakeClock, fakeRuntime, container, fakePodPullingTimeRecorder := pullerTestEnv(c, useSerializedEnv)
+               fakeRuntime.CalledFunctions = nil
+               fakeRuntime.ImageList = []Image{}
+               fakeClock.Step(time.Second)
                ctx := context.Background()
                _, _, err := puller.EnsureImageExists(ctx, pod, container, nil, nil)
                fakeRuntime.AssertCalls(c.expected[0].calls)
```

Reply:
Just to be sure our change does not introduce any problem in this area
<img width="472" alt="Screenshot 2023-01-10 at 11 39 04" src="https://user-images.githubusercontent.com/9576370/211529906-305191ff-4740-4646-9c64-42e5474f38ca.png">

Reply:
Done in 10639a7963b